src/coprocessor/dag/expr/arithmetic.rs

```diff
@@ -0,0 +1,390 @@
+
+use std::{f64, i64};
+use std::borrow::Cow;
+use coprocessor::codec::Datum;
+use coprocessor::codec::mysql::Decimal;
+use super::{Error, FnCall, Result, StatementContext};
+
+impl FnCall {
+    pub fn plus_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try!(self.children[0].eval_real(ctx, row));
+        let rhs = try!(self.children[1].eval_real(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = l + r;
+            if !res.is_finite() {
+                return Err(Error::Overflow);
+            }
+            Ok(res)
+        })
+    }
+
+    pub fn plus_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() + r.as_ref()).into();
+            result.map(Cow::Owned)
+        })
+    }
+
+    pub fn plus_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| l.checked_add(r).ok_or(Error::Overflow))
+    }
+
+    pub fn plus_uint(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = (l as u64).checked_add(r as u64).ok_or(Error::Overflow);
+            res.map(|u| u as i64)
+        })
+    }
+
+    pub fn minus_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try!(self.children[0].eval_real(ctx, row));
+        let rhs = try!(self.children[1].eval_real(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = l - r;
+            if !res.is_finite() {
+                return Err(Error::Overflow);
+            }
+            Ok(res)
+        })
+    }
+
+    pub fn minus_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() - r.as_ref()).into();
+            result.map(Cow::Owned)
+        })
+    }
+
+    pub fn minus_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| l.checked_sub(r).ok_or(Error::Overflow))
```

Comment:
TiDB converts both 2 arguments to unsigned, and tell TiKV use **MinusIntUnsigned** to calculate it. So `i64::MAX as u64 - (-1)`  won't cause any failure, but `u64::MAX - (-1)` will cause out-of-range failure.